File: Runtime/STSimTransformer.Patch.vb
'************************************************************************************
' ST-Sim: A .NET class library for developing state-and-transition simulation models.
'
' Copyright © 2009-2015 ApexRMS.
'
'************************************************************************************

Partial Class STSimTransformer

    ''' <summary>
    ''' Selects a patch initiation cell
    ''' </summary>
    ''' <param name="transitionGroup"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Private Function SelectPatchInitiationCell(ByVal transitionGroup As TransitionGroup) As Cell

        Debug.Assert(Me.IsSpatial)

        Dim pp As PatchPrioritization = transitionGroup.PatchPrioritization
        Dim patches As List(Of TransitionPatch) = pp.TransitionPatches

        If (patches.Count = 0) Then
            Return Nothing
        End If

        Dim Patch As TransitionPatch = patches(0)

        If (pp.PatchPrioritizationType = PatchPrioritizationType.Largest Or
            pp.PatchPrioritizationType = PatchPrioritizationType.LargestEdgesOnly) Then

            Patch = patches(patches.Count - 1)

        End If

        If (pp.PatchPrioritizationType = PatchPrioritizationType.SmallestEdgesOnly Or
            pp.PatchPrioritizationType = PatchPrioritizationType.LargestEdgesOnly) Then

            If (Patch.EdgeCells.ContainsKey(Patch.SeedCell.CellId)) Then
                Return Patch.SeedCell
            Else
                Return Patch.EdgeCells.Values.ElementAt(0)
            End If

        Else

            If (Patch.AllCells.ContainsKey(Patch.SeedCell.CellId)) Then
                Return Patch.SeedCell
            Else
                Return Patch.AllCells.Values.ElementAt(0)
            End If

        End If

    End Function

    ''' <summary>
    ''' Assigns patch prioritizations based on the current iteration and timestep
    ''' </summary>
    ''' <remarks></remarks>
    Private Sub AssignPatchPrioritizations(ByVal iteration As Integer, ByVal timestep As Integer)

        If (Not Me.IsSpatial) Then
            Return
        End If

        For Each tg As TransitionGroup In Me.m_TransitionGroups

            tg.PatchPrioritization = Nothing

            Dim tpp As TransitionPatchPrioritization =
                Me.m_TransitionPatchPrioritizationMap.GetPatchPrioritization(tg.TransitionGroupId, iteration, timestep)

            If (tpp IsNot Nothing) Then
                tg.PatchPrioritization = Me.m_PatchPrioritizations(tpp.PatchPrioritizationId)
            End If

        Next

    End Sub

    ''' <summary>
    ''' Determines if the specified cell is a edge cell in the specified patch
    ''' </summary>
    ''' <param name="cell"></param>
    ''' <param name="patch"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Private Function IsPatchEdgeCell(ByVal cell As Cell, ByVal patch As TransitionPatch) As Boolean

        If (Not CellInPatch(Me.GetCellNorth(cell), patch)) Then
            Return True
        End If

        If (Not CellInPatch(Me.GetCellNortheast(cell), patch)) Then
            Return True
        End If

        If (Not CellInPatch(Me.GetCellEast(cell), patch)) Then
            Return True
        End If

        If (Not CellInPatch(Me.GetCellSoutheast(cell), patch)) Then
            Return True
        End If

        If (Not CellInPatch(Me.GetCellSouth(cell), patch)) Then
            Return True
        End If

        If (Not CellInPatch(Me.GetCellSouthwest(cell), patch)) Then
            Return True
        End If

        If (Not CellInPatch(Me.GetCellWest(cell), patch)) Then
            Return True
        End If

        If (Not CellInPatch(Me.GetCellNorthwest(cell), patch)) Then
            Return True
        End If

        Return False

    End Function

    ''' <summary>
    ''' Determines if the specified cell is part of the specified patch
    ''' </summary>
    ''' <param name="cell"></param>
    ''' <param name="patch"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Private Shared Function CellInPatch(ByVal cell As Cell, ByVal patch As TransitionPatch) As Boolean

        If (cell Is Nothing) Then
            Return False
        End If

        If (patch.AllCells.ContainsKey(cell.CellId)) Then
            Return True
        End If

        Return False

    End Function

    ''' <summary>
    ''' Updates the patch membership for the specified transition group and cell
    ''' </summary>
    ''' <param name="transitionGroupId"></param>
    ''' <param name="cell"></param>
    ''' <remarks>
    ''' This function will also remove a patch if its cell collection becomes empty.
    ''' </remarks>
    Private Sub UpdateCellPatchMembership(ByVal transitionGroupId As Integer, ByVal cell As Cell)

        Dim TransitionGroup As TransitionGroup = Me.m_TransitionGroups(transitionGroupId)

        If (TransitionGroup.PatchPrioritization IsNot Nothing) Then

            Dim Patch As TransitionPatch = Nothing

            If (TransitionGroup.PatchPrioritization.PatchPrioritizationType = PatchPrioritizationType.Smallest Or
                TransitionGroup.PatchPrioritization.PatchPrioritizationType = PatchPrioritizationType.SmallestEdgesOnly) Then

                Patch = TransitionGroup.PatchPrioritization.TransitionPatches.First

            Else
                Patch = TransitionGroup.PatchPrioritization.TransitionPatches.Last
            End If

            If (Patch.AllCells.ContainsKey(cell.CellId)) Then
                Patch.AllCells.Remove(cell.CellId)
            End If

            If (Patch.EdgeCells.ContainsKey(cell.CellId)) Then
                Patch.EdgeCells.Remove(cell.CellId)
            End If

            If (TransitionGroup.PatchPrioritization.PatchPrioritizationType = PatchPrioritizationType.LargestEdgesOnly Or
                TransitionGroup.PatchPrioritization.PatchPrioritizationType = PatchPrioritizationType.SmallestEdgesOnly) Then

                If (Patch.EdgeCells.Count = 0) Then
                    TransitionGroup.PatchPrioritization.TransitionPatches.Remove(Patch)
                End If

            ElseIf (Patch.AllCells.Count = 0) Then

                Debug.Assert(Patch.EdgeCells.Count = 0)
                TransitionGroup.PatchPrioritization.TransitionPatches.Remove(Patch)

            End If

        End If

    End Sub

    ''' <summary>
    ''' Clears the transition patches for the specified transition group
    ''' </summary>
    ''' <param name="transitionGroup"></param>
    ''' <remarks></remarks>
    Private Sub ClearTransitionPatches(ByVal transitionGroup As TransitionGroup)

        Debug.Assert(Me.IsSpatial)

        If (transitionGroup.PatchPrioritization IsNot Nothing) Then
            transitionGroup.PatchPrioritization.TransitionPatches.Clear()
        End If

    End Sub

    ''' <summary>
    ''' Grows a transition patch
    ''' </summary>
    ''' <param name="transitionedCells"></param>
    ''' <param name="patchCells"></param>
    ''' <param name="initiationCell"></param>
    ''' <param name="patch"></param>
    ''' <param name="transitionGroup"></param>
    ''' <param name="iteration"></param>
    ''' <param name="timestep"></param>
    ''' <remarks></remarks>
    Private Sub GrowTransitionPatch(
        ByVal transitionedCells As Dictionary(Of Integer, Cell),
        ByVal patchCells As Dictionary(Of Integer, Cell),
        ByVal initiationCell As Cell,
        ByVal patch As TransitionPatch,
        ByVal transitionGroup As TransitionGroup,
        ByVal iteration As Integer,
        ByVal timestep As Integer)

        Dim PatchSize As Double = 0.0
        Dim PatchQueue As New Queue(Of Cell)
        Dim QueuedCells As New Dictionary(Of Integer, Cell)

        PatchQueue.Enqueue(initiationCell)
        QueuedCells.Add(initiationCell.CellId, initiationCell)

        While (PatchQueue.Count > 0)

            Dim CurrentCell As Cell = PatchQueue.Dequeue()

            Debug.Assert(Not patchCells.ContainsKey(CurrentCell.CellId))
            Debug.Assert(Not patch.AllCells.ContainsKey(CurrentCell.CellId))

            patchCells.Add(CurrentCell.CellId, CurrentCell)
            patch.AllCells.Add(CurrentCell.CellId, CurrentCell)

            PatchSize += Me.m_AmountPerCell

            Me.AddNeighboringPatchCell(transitionedCells, patchCells, QueuedCells, PatchQueue, Me.GetCellNorth(CurrentCell), transitionGroup, iteration, timestep)
            Me.AddNeighboringPatchCell(transitionedCells, patchCells, QueuedCells, PatchQueue, Me.GetCellNortheast(CurrentCell), transitionGroup, iteration, timestep)
            Me.AddNeighboringPatchCell(transitionedCells, patchCells, QueuedCells, PatchQueue, Me.GetCellEast(CurrentCell), transitionGroup, iteration, timestep)
            Me.AddNeighboringPatchCell(transitionedCells, patchCells, QueuedCells, PatchQueue, Me.GetCellSoutheast(CurrentCell), transitionGroup, iteration, timestep)
            Me.AddNeighboringPatchCell(transitionedCells, patchCells, QueuedCells, PatchQueue, Me.GetCellSouth(CurrentCell), transitionGroup, iteration, timestep)
            Me.AddNeighboringPatchCell(transitionedCells, patchCells, QueuedCells, PatchQueue, Me.GetCellSouthwest(CurrentCell), transitionGroup, iteration, timestep)
            Me.AddNeighboringPatchCell(transitionedCells, patchCells, QueuedCells, PatchQueue, Me.GetCellWest(CurrentCell), transitionGroup, iteration, timestep)
            Me.AddNeighboringPatchCell(transitionedCells, patchCells, QueuedCells, PatchQueue, Me.GetCellNorthwest(CurrentCell), transitionGroup, iteration, timestep)

        End While

        patch.Size = PatchSize

        For Each PatchCell In patch.AllCells.Values

            If (Me.IsPatchEdgeCell(PatchCell, patch)) Then
                patch.EdgeCells.Add(PatchCell.CellId, PatchCell)
            End If

        Next

        Debug.Assert(patch.EdgeCells.Count > 0)
        Debug.Assert(patch.AllCells.Count > 0)
        Debug.Assert(patch.AllCells.Count >= patch.EdgeCells.Count)

    End Sub

    ''' <summary>
    ''' Adds a neighboring patch cell
    ''' </summary>
    ''' <param name="transitionedCells"></param>
    ''' <param name="patchCells"></param>
    ''' <param name="queuedCells"></param>
    ''' <param name="patchQueue"></param>
    ''' <param name="neighborCell"></param>
    ''' <param name="transitionGroup"></param>
    ''' <param name="iteration"></param>
    ''' <param name="timestep"></param>
    ''' <remarks></remarks>
    Private Sub AddNeighboringPatchCell(
        ByVal transitionedCells As Dictionary(Of Integer, Cell),
        ByVal patchCells As Dictionary(Of Integer, Cell),
        ByVal queuedCells As Dictionary(Of Integer, Cell),
        ByVal patchQueue As Queue(Of Cell),
        ByVal neighborCell As Cell,
        ByVal transitionGroup As TransitionGroup,
        ByVal iteration As Integer,
        ByVal timestep As Integer)

        If (neighborCell IsNot Nothing) Then

            If (queuedCells.ContainsKey(neighborCell.CellId) Or
                patchCells.ContainsKey(neighborCell.CellId) Or
                transitionedCells.ContainsKey(neighborCell.CellId)) Then

                Return

            End If

            If (Me.SelectSpatialTransitionPathway(
                neighborCell,
                transitionGroup.TransitionGroupId,
                iteration,
                timestep) Is Nothing) Then

                Return

            End If

            queuedCells.Add(neighborCell.CellId, neighborCell)
            patchQueue.Enqueue(neighborCell)

        End If

    End Sub

    ''' <summary>
    ''' Fills the transition patches for the specified transition group
    ''' </summary>
    ''' <param name="transitionedCells"></param>
    ''' <param name="stratum"></param>
    ''' <param name="transitionGroup"></param>
    ''' <param name="iteration"></param>
    ''' <param name="timestep"></param>
    ''' <remarks></remarks>
    Private Sub FillTransitionPatches(
        ByVal transitionedCells As Dictionary(Of Integer, Cell),
        ByVal stratum As Stratum,
        ByVal transitionGroup As TransitionGroup,
        ByVal iteration As Integer,
        ByVal timestep As Integer)

        Debug.Assert(Me.IsSpatial)

        If (transitionGroup.PatchPrioritization Is Nothing) Then
            Return
        End If

        Debug.Assert(transitionGroup.PatchPrioritization.TransitionPatches.Count = 0)

        Dim patchCells As New Dictionary(Of Integer, Cell)

        For Each SimulationCell As Cell In stratum.Cells.Values

            Debug.Assert(SimulationCell.StratumId <> 0)
            Debug.Assert(SimulationCell.StateClassId <> 0)

            If (patchCells.ContainsKey(SimulationCell.CellId) Or
                transitionedCells.ContainsKey(SimulationCell.CellId)) Then

                Continue For

            End If

            If (Me.SelectSpatialTransitionPathway(
                SimulationCell,
                transitionGroup.TransitionGroupId,
                iteration,
                timestep) Is Nothing) Then

                Continue For

            End If

            Dim Patch As New TransitionPatch(SimulationCell)

            Me.GrowTransitionPatch(
                transitionedCells,
                patchCells,
                SimulationCell,
                Patch,
                transitionGroup,
                iteration,
                timestep)

            transitionGroup.PatchPrioritization.TransitionPatches.Add(Patch)

        Next

        transitionGroup.PatchPrioritization.TransitionPatches.Sort(
            Function(p1 As TransitionPatch, p2 As TransitionPatch) As Integer
                Return (p1.Size.CompareTo(p2.Size))
            End Function)

    End Sub

End Class

File: Runtime/STSimTransformer.Patch.cs
// ST-Sim: A SyncroSim Module for the ST-Sim State-and-Transition Model.
// Copyright © 2007-2018 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.

using System.Diagnostics;
using System.Collections.Generic;
using System.Linq;

namespace SyncroSim.STSim
{
    public partial class STSimTransformer
    {
        /// <summary>
        /// Selects a patch initiation cell
        /// </summary>
        /// <param name="transitionGroup"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        private Cell SelectPatchInitiationCell(TransitionGroup transitionGroup)
        {
            Debug.Assert(this.IsSpatial);

            PatchPrioritization pp = transitionGroup.PatchPrioritization;
            List<TransitionPatch> patches = pp.TransitionPatches;

            if (patches.Count == 0)
            {
                return null;
            }

            TransitionPatch Patch = patches[0];

            if (pp.PatchPrioritizationType == PatchPrioritizationType.Largest ||
                pp.PatchPrioritizationType == PatchPrioritizationType.LargestEdgesOnly)
            {
                Patch = patches[patches.Count - 1];
            }

            if (pp.PatchPrioritizationType == PatchPrioritizationType.SmallestEdgesOnly || 
                pp.PatchPrioritizationType == PatchPrioritizationType.LargestEdgesOnly)
            {
                if (Patch.EdgeCells.ContainsKey(Patch.SeedCell.CellId))
                {
                    return Patch.SeedCell;
                }
                else
                {
                    return Patch.EdgeCells.Values.ElementAt(0);
                }
            }
            else
            {
                if (Patch.AllCells.ContainsKey(Patch.SeedCell.CellId))
                {
                    return Patch.SeedCell;
                }
                else
                {
                    return Patch.AllCells.Values.ElementAt(0);
                }
            }
        }

        /// <summary>
        /// Assigns patch prioritizations based on the current iteration and timestep
        /// </summary>
        /// <remarks></remarks>
        private void AssignPatchPrioritizations(int iteration, int timestep)
        {
            if (!this.IsSpatial)
            {
                return;
            }

            foreach (TransitionGroup tg in this.m_TransitionGroups)
            {
                tg.PatchPrioritization = null;

                TransitionPatchPrioritization tpp = this.m_TransitionPatchPrioritizationMap.GetPatchPrioritization(
                    tg.TransitionGroupId, iteration, timestep);

                if (tpp != null)
                {
                    tg.PatchPrioritization = this.m_PatchPrioritizations[tpp.PatchPrioritizationId];
                }
            }
        }

        /// <summary>
        /// Determines if the specified cell is a edge cell in the specified patch
        /// </summary>
        /// <param name="cell"></param>
        /// <param name="patch"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        private bool IsPatchEdgeCell(Cell cell, TransitionPatch patch)
        {
            if (!CellInPatch(this.GetCellNorth(cell), patch))
            {
                return true;
            }

            if (!CellInPatch(this.GetCellNortheast(cell), patch))
            {
                return true;
            }

            if (!CellInPatch(this.GetCellEast(cell), patch))
            {
                return true;
            }

            if (!CellInPatch(this.GetCellSoutheast(cell), patch))
            {
                return true;
            }

            if (!CellInPatch(this.GetCellSouth(cell), patch))
            {
                return true;
            }

            if (!CellInPatch(this.GetCellSouthwest(cell), patch))
            {
                return true;
            }

            if (!CellInPatch(this.GetCellWest(cell), patch))
            {
                return true;
            }

            if (!CellInPatch(this.GetCellNorthwest(cell), patch))
            {
                return true;
            }

            return false;
        }

        /// <summary>
        /// Determines if the specified cell is part of the specified patch
        /// </summary>
        /// <param name="cell"></param>
        /// <param name="patch"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        private static bool CellInPatch(Cell cell, TransitionPatch patch)
        {
            if (cell == null)
            {
                return false;
            }

            if (patch.AllCells.ContainsKey(cell.CellId))
            {
                return true;
            }

            return false;
        }

        /// <summary>
        /// Updates the patch membership for the specified transition group and cell
        /// </summary>
        /// <param name="transitionGroupId"></param>
        /// <param name="cell"></param>
        /// <remarks>
        /// This function will also remove a patch if its cell collection becomes empty.
        /// </remarks>
        private void UpdateCellPatchMembership(int transitionGroupId, Cell cell)
        {
            TransitionGroup TransitionGroup = this.m_TransitionGroups[transitionGroupId];

            if (TransitionGroup.PatchPrioritization != null)
            {
                TransitionPatch Patch = null;

                if (
                    TransitionGroup.PatchPrioritization.PatchPrioritizationType == PatchPrioritizationType.Smallest || 
                    TransitionGroup.PatchPrioritization.PatchPrioritizationType == PatchPrioritizationType.SmallestEdgesOnly)
                {
                    Patch = TransitionGroup.PatchPrioritization.TransitionPatches.First();
                }
                else
                {
                    Patch = TransitionGroup.PatchPrioritization.TransitionPatches.Last();
                }

                if (Patch.AllCells.ContainsKey(cell.CellId))
                {
                    Patch.AllCells.Remove(cell.CellId);
                }

                if (Patch.EdgeCells.ContainsKey(cell.CellId))
                {
                    Patch.EdgeCells.Remove(cell.CellId);
                }

                if (
                    TransitionGroup.PatchPrioritization.PatchPrioritizationType == PatchPrioritizationType.LargestEdgesOnly || 
                    TransitionGroup.PatchPrioritization.PatchPrioritizationType == PatchPrioritizationType.SmallestEdgesOnly)
                {
                    if (Patch.EdgeCells.Count() == 0)
                    {
                        TransitionGroup.PatchPrioritization.TransitionPatches.Remove(Patch);
                    }
                }
                else if (Patch.AllCells.Count() == 0)
                {
                    Debug.Assert(Patch.EdgeCells.Count() == 0);
                    TransitionGroup.PatchPrioritization.TransitionPatches.Remove(Patch);
                }
            }
        }

        /// <summary>
        /// Clears the transition patches for the specified transition group
        /// </summary>
        /// <param name="transitionGroup"></param>
        /// <remarks></remarks>
        private void ClearTransitionPatches(TransitionGroup transitionGroup)
        {
            Debug.Assert(this.IsSpatial);

            if (transitionGroup.PatchPrioritization != null)
            {
                transitionGroup.PatchPrioritization.TransitionPatches.Clear();
            }
        }

        /// <summary>
        /// Grows a transition patch
        /// </summary>
        /// <param name="transitionedCells"></param>
        /// <param name="patchCells"></param>
        /// <param name="initiationCell"></param>
        /// <param name="patch"></param>
        /// <param name="transitionGroup"></param>
        /// <param name="iteration"></param>
        /// <param name="timestep"></param>
        /// <remarks></remarks>
        private void GrowTransitionPatch(
            Dictionary<int, Cell> transitionedCells, Dictionary<int, Cell> patchCells, Cell initiationCell, 
            TransitionPatch patch, TransitionGroup transitionGroup, int iteration, int timestep)
        {
            double PatchSize = 0.0;
            Queue<Cell> PatchQueue = new Queue<Cell>();
            Dictionary<int, Cell> QueuedCells = new Dictionary<int, Cell>();

            PatchQueue.Enqueue(initiationCell);
            QueuedCells.Add(initiationCell.CellId, initiationCell);

            while (PatchQueue.Count > 0)
            {
                Cell CurrentCell = PatchQueue.Dequeue();

                Debug.Assert(!patchCells.ContainsKey(CurrentCell.CellId));
                Debug.Assert(!patch.AllCells.ContainsKey(CurrentCell.CellId));

                patchCells.Add(CurrentCell.CellId, CurrentCell);
                patch.AllCells.Add(CurrentCell.CellId, CurrentCell);

                PatchSize += this.m_AmountPerCell;

                this.AddNeighboringPatchCell(transitionedCells, patchCells, QueuedCells, PatchQueue, this.GetCellNorth(CurrentCell), transitionGroup, iteration, timestep);
                this.AddNeighboringPatchCell(transitionedCells, patchCells, QueuedCells, PatchQueue, this.GetCellNortheast(CurrentCell), transitionGroup, iteration, timestep);
                this.AddNeighboringPatchCell(transitionedCells, patchCells, QueuedCells, PatchQueue, this.GetCellEast(CurrentCell), transitionGroup, iteration, timestep);
                this.AddNeighboringPatchCell(transitionedCells, patchCells, QueuedCells, PatchQueue, this.GetCellSoutheast(CurrentCell), transitionGroup, iteration, timestep);
                this.AddNeighboringPatchCell(transitionedCells, patchCells, QueuedCells, PatchQueue, this.GetCellSouth(CurrentCell), transitionGroup, iteration, timestep);
                this.AddNeighboringPatchCell(transitionedCells, patchCells, QueuedCells, PatchQueue, this.GetCellSouthwest(CurrentCell), transitionGroup, iteration, timestep);
                this.AddNeighboringPatchCell(transitionedCells, patchCells, QueuedCells, PatchQueue, this.GetCellWest(CurrentCell), transitionGroup, iteration, timestep);
                this.AddNeighboringPatchCell(transitionedCells, patchCells, QueuedCells, PatchQueue, this.GetCellNorthwest(CurrentCell), transitionGroup, iteration, timestep);
            }

            patch.Size = PatchSize;

            foreach (var PatchCell in patch.AllCells.Values)
            {
                if (this.IsPatchEdgeCell(PatchCell, patch))
                {
                    patch.EdgeCells.Add(PatchCell.CellId, PatchCell);
                }
            }

            Debug.Assert(patch.EdgeCells.Count() > 0);
            Debug.Assert(patch.AllCells.Count() > 0);
            Debug.Assert(patch.AllCells.Count() >= patch.EdgeCells.Count());
        }

        /// <summary>
        /// Adds a neighboring patch cell
        /// </summary>
        /// <param name="transitionedCells"></param>
        /// <param name="patchCells"></param>
        /// <param name="queuedCells"></param>
        /// <param name="patchQueue"></param>
        /// <param name="neighborCell"></param>
        /// <param name="transitionGroup"></param>
        /// <param name="iteration"></param>
        /// <param name="timestep"></param>
        /// <remarks></remarks>
        private void AddNeighboringPatchCell
            (Dictionary<int, Cell> transitionedCells, Dictionary<int, Cell> patchCells, Dictionary<int, Cell> queuedCells, 
            Queue<Cell> patchQueue, Cell neighborCell, TransitionGroup transitionGroup, int iteration, int timestep)
        {
            if (neighborCell != null)
            {
                if (
                    queuedCells.ContainsKey(neighborCell.CellId) || 
                    patchCells.ContainsKey(neighborCell.CellId) || 
                    transitionedCells.ContainsKey(neighborCell.CellId))
                {
                    return;
                }

                if (this.SelectSpatialTransitionPathway(neighborCell, transitionGroup.TransitionGroupId, iteration, timestep) == null)
                {
                    return;
                }

                queuedCells.Add(neighborCell.CellId, neighborCell);
                patchQueue.Enqueue(neighborCell);
            }
        }

        /// <summary>
        /// Fills the transition patches for the specified transition group
        /// </summary>
        /// <param name="transitionedCells"></param>
        /// <param name="stratum"></param>
        /// <param name="transitionGroup"></param>
        /// <param name="iteration"></param>
        /// <param name="timestep"></param>
        /// <remarks></remarks>
        private void FillTransitionPatches(
            Dictionary<int, Cell> transitionedCells, Stratum stratum, 
            TransitionGroup transitionGroup, int iteration, int timestep)
        {
            Debug.Assert(this.IsSpatial);

            if (transitionGroup.PatchPrioritization == null)
            {
                return;
            }

            Debug.Assert(transitionGroup.PatchPrioritization.TransitionPatches.Count == 0);

            Dictionary<int, Cell> patchCells = new Dictionary<int, Cell>();

            foreach (Cell SimulationCell in stratum.Cells.Values)
            {
                Debug.Assert(SimulationCell.StratumId != 0);
                Debug.Assert(SimulationCell.StateClassId != 0);

                if (patchCells.ContainsKey(SimulationCell.CellId) | transitionedCells.ContainsKey(SimulationCell.CellId))
                {
                    continue;
                }

                if (this.SelectSpatialTransitionPathway(SimulationCell, transitionGroup.TransitionGroupId, iteration, timestep) == null)
                {
                    continue;
                }

                TransitionPatch Patch = new TransitionPatch(SimulationCell);

                this.GrowTransitionPatch(transitionedCells, patchCells, SimulationCell, Patch, transitionGroup, iteration, timestep);

                transitionGroup.PatchPrioritization.TransitionPatches.Add(Patch);
            }

            transitionGroup.PatchPrioritization.TransitionPatches.Sort((TransitionPatch p1, TransitionPatch p2) =>
            {
                return (p1.Size.CompareTo(p2.Size));
            });
        }
    }
}
